ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -577,7 +543,14 @@ private void visitSelectStar(final AllColumns allColumns) {
             : "";
 
         final LogicalSchema schema = source.getDataSource().getSchema();
-        for (final Column column : schema.columns()) {
+
+        // Non-join persistent queries only require value columns on SELECT *
```

Comment:
What should the key schema be in this query after the changes to `PARTITION BY`? Will we support aliases in `PARTITION BY` columns?

Reply:
> EDIT: #3731 - can you add this ticket to this inline comment?

No, because this PR does not exclude `ROWKEY` from the join schema yet.  It will still have `LEFT.ROWKEY` and `RIGHT.ROWKEY`.  

What this PR is doing is changing the rules in `Analyzer` so that:

a) explicit un-aliased `ROWKEY`, (and `ROWTIME`), columns are rejected in C*AS statements, as they would be removed by the `KsqlStructuredDataOutputNode` node anyway. (See changes to the QTT, where I've changed tests that I added to prove the weird behaviour to now reject the statement).
b) Not expand `SELECT *` for non-join persistent queries to include  `ROWKEY` and `ROWTIME` (as they are not allowed in the projection, as per point a above.

Reply:
> what about queries that have group-by or repartitions? I forgot what we decided about a `*` copying key fields, but are we suggesting that:
> 
> ```sql
> SELECT * FROM foo PARTITION BY substring(col1, 2);
> ```
> 
> would not have the key field from `foo` accessible? 

This PR doesn't change anything to do with how columns are available in a join or partition by.  ROWKEY will still be accessible, it just needs to be aliased, e.g.  there is a QTT test in `partition-by.json` that does:

```json
{
      "name": "partition by project ROWKEY",
      "statements": [
        "CREATE STREAM INPUT (ID bigint) with (kafka_topic='input', value_format='JSON');",
        "CREATE STREAM OUTPUT AS select ROWKEY AS OLDKEY from INPUT partition by ID;"
      ],
      "inputs": [{"topic": "input", "key": "foo", "value": {"ID":  10}}],
      "outputs": [{"topic": "OUTPUT", "key": "10", "value": {"OLDKEY": "foo"}}]
    }
```

> If that's the case, then why are joins any different?

Joins are different because the left and right `ROWKEY` columns on a `SELECT *` come through aliased, so aren't auto-removed later.  #3731 proposes we should remove them, but that's another PR.

Reply:
> What should the key schema be in this query after the changes to `PARTITION BY`? Will we support aliases in `PARTITION BY` columns?

Not sure I follow the question.  Can you elaborate? Does the above answers cover this?

To be clear through, the change here is just tightening up how `SELECT *` is expanded to not include select columns that will ultimately be removed later in the query and shouldn't be allowed in the query in the first place.

`ROWKEY` is still copied into the value by the KS implementation. `ROWKEY` is still available to be aliased in the projection as needed, etc,

Reply:
> > EDIT: #3731 - can you add this ticket to this inline comment?
> 
> No, because this PR does not exclude `ROWKEY` from the join schema yet. It will still have `LEFT.ROWKEY` and `RIGHT.ROWKEY`.
> 
> What this PR is doing is changing the rules in `Analyzer` so that:
> 
> a) explicit un-aliased `ROWKEY`, (and `ROWTIME`), columns are rejected in C*AS statements, as they would be removed by the `KsqlStructuredDataOutputNode` node anyway. (See changes to the QTT, where I've changed tests that I added to prove the weird behaviour to now reject the statement).
> b) Not expand `SELECT *` for non-join persistent queries to include `ROWKEY` and `ROWTIME` (as they are not allowed in the projection, as per point a above.

Sorry I just realized how confusing my comment might be. After doing more reading, I realized why we specify "non-join persistent queries" in the inline comment - it's because we need to implement #3731. All I was asking for in my comment is to add a reference to that ticket to the inline comment so that other people reading this code will understand why :)